### PR TITLE
refactor: poll for git auth updates when creating a workspace

### LINKED ISSUE #9536: Git auth to create a workspace forces page refresh losing all fields
Problem:
- When a user is creating a workspace with a template that requires git auth to some external provider a user is forced to refresh the page for the auth to take effect which removes all entries in other fields when creating the workspace.  This is a problem because it is not apparent that all fields will be reset leading to end user frustration. 

Solutions:
1. (Preferred) Turn git auth buttons and settings into a websocket or something that is updated dynamically as soon as the end user completes the git auth which will negate the need for a full page refresh.
2. Enforce git auth on a separate page or first before other fields are filled which negates the removal of selections and filled fields
3. More I didn't list

### PR DESCRIPTION
**Pull request by aslilac**
Closes #9536

- Move `gitAuth` management out of xstate
- Poll for updates for one minute (should be more than plenty) after clicking the "Login with [x]" button
- Show a "check again" button in case one minute isn't long enough
- Remove some dangling `BroadcastChannel` stuff that wasn't working (and didn't even seem fully implemented)

### CODE REVIEW

**Comment by aslilac** [comment]
Remaining improvements I wanna do before merging:

- [ ] Stop polling when all auth providers are connected
- [ ] I think there might've been more `BroadcastChannel` stuff I can remove

Shouldn't be hard, but I wanted to get this PR up before I quit for the day

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx
Comment (BrunoQuaresma) [inline site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:62]: I think you can move the polling to react-query using `fetchInterval` API like:
```ts
// Pseudo code
useQuery({
   fetchInterval: (data) => needsPooling(data) ? 2_000 : false
})
```

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx
Comment (BrunoQuaresma) [inline site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:90]: Ahhh I see you are doing it here... wondering why you need an extra state for the pulling 🤔

**Comment by BrunoQuaresma** [comment]
I will wait for the remaining work before approve it 👍

**Comment by BrunoQuaresma** [comment]
Also I'm missing tests to avoid any kind of regressions, do you think we could add some coverage to it?

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx
Comment (Parkreiner) [inline site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:113]: @BrunoQuaresma Still haven't had the chance to look through XState, so I have no idea here: is XState just not able to make more type-safe guarantees? I'm surprised that it requires all these `undefined` checks

File: site/src/pages/GitAuthPage/GitAuthPageView.tsx
Comment (Parkreiner) [inline site/src/pages/GitAuthPage/GitAuthPageView.tsx:79]: Nitpick, but should there be formatting for when the login isn't available for whatever reason? If `gitAuth.user?.login` evaluates to `undefined`, then the user would see `Hey @! 👋`. Maybe a generic `Hey there!` would be good?

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx
Comment (BrunoQuaresma) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:113]: It is something they have to improve for sure. It is not 100% type-safe in v4 but looks like they are looking to improve it in v5.

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx
Comment (aslilac) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:84]: that's not the intended behavior. the idea is to wait until we know there's a chance that something is worth polling.

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx
Comment (aslilac) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:119]: that's a more complicated change that it might seem. I don't actually know what the types this will have at runtime are, or if there's some weird loading logic. this was just a nice quick change to a type assertion that makes the intent more clear, rather than a complete cast that might do something nasty with a very low chance of throwing a compiler error.

File: site/src/pages/GitAuthPage/GitAuthPageView.tsx
Comment (aslilac) [reply site/src/pages/GitAuthPage/GitAuthPageView.tsx:79]: I just don't wanna creep the scope of this PR too much with things that _might_ be wrong. I was just correcting something that _definitely_ looked wrong to me. 😅 idk if that seems like a silly difference, but if I chased every weird thing I noticed I'd never finish any PRs.

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx
Comment (Parkreiner) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:84]: I feel like I said this weird. What I was trying to say is that if the goal is to make sure that React does nothing at all in the off chance that the callback is called while polling is already happening, `setGitAuthPollingState("polling")` will do that.

The function will still dispatch a new value, but because it would be the same value as before (especially with the state being a primitive), React rejects the state change and doesn't re-render

File: site/src/pages/GitAuthPage/GitAuthPageView.tsx
Comment (Parkreiner) [reply site/src/pages/GitAuthPage/GitAuthPageView.tsx:79]: No, that's fair, and that's a perspective I really should adopt more myself

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx
Comment (aslilac) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:84]: oh, sorry 😅 I think you're right

File: site/src/xServices/createWorkspace/createWorkspaceXService.ts
Comment (aslilac) [reply site/src/xServices/createWorkspace/createWorkspaceXService.ts:253]: This is a little different than that case, because it has a timeout where it will stop polling until the user interacts again, so this definitely won't be an "increases load by xx%" type of thing. I get the appeal of `BroadcastChannel`, but I've always found that it ends up being fragile for stuff like this.

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx
Comment (aslilac) [reply site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:90]: the third state shows the retry button

**Review by BrunoQuaresma — APPROVED** [review]

**Comment by aslilac** [comment]
I've spent a bit of time trying to figure out how the `Broadcast` channel was supposed to work, and it seems like a serious chunk of it is missing, _and_ like it might've had a couple bugs anyway. (Namely, it didn't include any way to tell which `GitAuth` was completed. I can't tell if that triggered a refetch or if it all of the auth providers were supposed to assume the message was for them.)

There's also some other good cleanup stuff in this PR that I want to get merged, so I'm just gonna go with this as-is since it's tested and working. If we want to revisit `BroadcastChannel` stuff in the future, we can schedule some time for me to circle back.

**Pull request merged**
